File: scripts/check_honest_kernel_bind.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""Fail-closed contract: kernel chips bind /honest locked_formula_count."""

from __future__ import annotations

import re
from pathlib import Path


ROOT = Path(__file__).resolve().parents[1]
INDEX = ROOT / "index.html"
DILIGENCE = ROOT / "diligence" / "index.html"
BIND = ROOT / "scripts" / "honest_kernel_bind.js"
HEADERS = ROOT / "_headers"
HONEST_URL = "https://a-11-oy.com/api/a11oy/v1/honest"
HONEST_FIELD = "locked_formula_count"
HTML_SURFACES = tuple(
    path
    for path in sorted(ROOT.rglob("*.html"))
    if ".git" not in path.parts
)
KERNEL_CHIP = re.compile(
    r"""id=["']cnt-locked["']"""
    r"""|id=["']pt-locked["']"""
    r"""|id=["']hs-proven["']"""
    r"""|data-kernel-chip"""
    r"""|locked-proven\s*=""",
    re.I,
)
HONEST_BIND = re.compile(
    r"""(?:/api/a11oy/v1/honest|/honest)[\s\S]{0,500}?locked_formula_count"""
    r"""|locked_formula_count[\s\S]{0,240}?(?:/api/a11oy/v1/honest|/honest)""",
    re.I,
)
HARDCODED_EIGHT = re.compile(
    r"""locked-proven\s*=\s*(?:<b>)?exactly\s*8"""
    r"""|locked-proven\s*=\s*exactly\s+eight""",
    re.I,
)
CATALOG_LABEL = re.compile(
    r"""(?:genome|catalog)\s+LOCKED-PROVEN|LOCKED-PROVEN\s+catalog""",
    re.I,
)


def neutral_semantic_style(source: str, selector: str) -> bool:
    """Require semantic chips to remain neutral, never proof/live colored.

    The selector may be qualified by a component scope (for example
    ``.hero-canon .catalog-chip``). We validate the declaration semantics
    rather than a historical minifier byte sequence.
    """
    escaped = re.escape(selector)
    patterns = (
        rf"{escaped}[^{{]*\{{[^}}]*color\s*:\s*var\(--gray\)",
        rf"{escaped}[^{{]*\{{[^}}]*color\s*:\s*var\(--t2\)",
    )
    return any(re.search(pattern, source, re.I) for pattern in patterns)


def check() -> None:
    bind = BIND.read_text(encoding="utf-8")
    assert BIND.is_file()
    assert HONEST_URL in bind
    assert HONEST_FIELD in bind
    assert "value === 8" in bind
    assert "N/A" in bind
    assert "UNAVAILABLE" in bind
    assert re.search(r"\?\?\s*8\b", bind) is None
    assert re.search(r"\|\|\s*8\b", bind) is None
    assert "exactly 8" not in bind.lower()
    assert "exactly eight" not in bind.lower()

    index = INDEX.read_text(encoding="utf-8")
    assert HARDCODED_EIGHT.search(index) is None, (
        "hero/footer kernel chips must not paint a hardcoded exactly 8"
    )
    assert 'data-kernel-chip="locked-proven"' in index
    assert f'data-honest-url="{HONEST_URL}"' in index
    assert f'data-honest-field="{HONEST_FIELD}"' in index
    assert 'id="cnt-locked"' in index
    assert ">N/A<" in index
    assert HONEST_BIND.search(index)
    assert 'src="scripts/honest_kernel_bind.js"' in index
    assert "https://a-11-oy.com" in index
    assert CATALOG_LABEL.search(index), "catalog LOCKED-PROVEN=25 must stay labelled"
    assert "catalog LOCKED-PROVEN" in index
    catalog = re.search(
        r'class="catalog-chip"[^>]*>(.*?)</span>',
        index,
        re.I | re.S,
    )
    assert catalog, "catalog 25 must be a catalog-chip, not the kernel"
    catalog_html = catalog.group(0)
    assert "25" in catalog_html
    assert "glass" not in catalog_html
    assert "record" not in catalog_html
    assert "live" not in catalog_html
    assert "var(--proof)" not in catalog_html
    catalog_text = catalog.group(1).lower()
    assert "genome catalog" in catalog_text or "not the kernel" in catalog_text
    assert "Lean-8 ≠ genome-144" in index or "Lean-8 != genome-144" in index
    assert 'class="frontier-kernel"' in index
    assert "not ROADMAP" in index
    assert 'class="conjecture"' in index
    assert "Conjecture 1" in index

    kanchay = (ROOT / "assets" / "kanchay.css").read_text(encoding="utf-8")
    combined_style = index + "\n" + kanchay
    assert neutral_semantic_style(combined_style, ".conjecture"), (
        "Conjecture must remain hierarchy-neutral (gray/t2), never proof/live colored"
    )
    assert neutral_semantic_style(combined_style, ".catalog-chip"), (
        "catalog chip must remain hierarchy-neutral (gray/t2), never proof/live colored"
    )
    for semantic_class in ("conjecture", "catalog-chip"):
        assert re.search(
            rf"\.{semantic_class}[^{{]*\{{[^}}]*color\s*:\s*var\(--proof\)",
            combined_style,
            re.I,
        ) is None, f"{semantic_class} must never inherit proof/live color"

    kernel_hosts = re.findall(
        r"<span[^>]*data-kernel-chip[^>]*>.*?</span>",
        index,
        re.I | re.S,
    )
    assert kernel_hosts, "kernel chips must exist so the bind can be probed"
    for host in kernel_hosts:
        assert "LOCKED-PROVEN" not in host
        assert "exactly 8" not in host.lower()
        assert ">25<" not in host

    headers = HEADERS.read_text(encoding="utf-8")
    assert "connect-src 'self' https://huggingface.co https://a-11-oy.com" in headers
    meta = re.search(
        r'http-equiv="Content-Security-Policy" content="([^"]+)"',
        index,
    )
    assert meta, "root meta CSP required"
    assert "connect-src 'self' https://huggingface.co https://a-11-oy.com" in meta.group(1)

    saw_chip = False
    for path in HTML_SURFACES:
        text = path.read_text(encoding="utf-8")
        relative = str(path.relative_to(ROOT))
        assert HARDCODED_EIGHT.search(text) is None, f"{relative}: hardcoded kernel 8"
        has_chip = bool(KERNEL_CHIP.search(text))
        if has_chip:
            saw_chip = True
            assert HONEST_BIND.search(text), (
                f"{relative}: kernel chip must bind {HONEST_URL} {HONEST_FIELD}"
            )
            if path == DILIGENCE:
                assert "N/A" in text
                assert CATALOG_LABEL.search(text)
                assert "Lean-8 ≠ genome-144" in text or "Lean-8 != genome-144" in text
                assert "Conjecture 1" in text
    assert saw_chip, "missing kernel-chip probe is FAIL"


if __name__ == "__main__":
    check()
    print(
        "OK: kernel chips bind /honest locked_formula_count === 8 or N/A; "
        "catalog 25 stays labelled."
    )

File: scripts/check_honest_kernel_bind.mjs
// SPDX-License-Identifier: Apache-2.0
import assert from "node:assert/strict";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const policy = require("./honest_kernel_bind.js");

assert.equal(policy.HONEST_URL, "https://a-11-oy.com/api/a11oy/v1/honest");
assert.equal(policy.HONEST_FIELD, "locked_formula_count");

assert.equal(policy.lockedFormulaCountFromHonest({ locked_formula_count: 8 }), 8);
assert.equal(
  policy.lockedFormulaCountFromHonest({
    doctrine_lock: { locked_formula_count: 8 },
  }),
  8,
);
assert.equal(policy.lockedFormulaCountFromHonest({ kernel_commit: "c7c0ba17" }), null);
assert.equal(policy.lockedFormulaCountFromHonest({ locked_formula_count: 25 }), 25);
assert.equal(policy.lockedFormulaCountFromHonest({ locked_formula_count: "8" }), null);
assert.equal(policy.lockedFormulaCountFromHonest(null), null);

assert.equal(policy.isExactLockedKernelCount(8), true);
assert.equal(policy.isExactLockedKernelCount(7), false);
assert.equal(policy.isExactLockedKernelCount(25), false);
assert.equal(policy.isExactLockedKernelCount(null), false);

assert.equal(policy.kernelCountLabel(8), "8");
assert.equal(policy.kernelCountLabel(7), "N/A");
assert.equal(policy.kernelCountLabel(25), "N/A");
assert.equal(policy.kernelCountLabel(144), "N/A");
assert.equal(policy.kernelCountLabel(null), "N/A");

const observed = policy.classifyHonestPayload({
  doctrine_lock: {
    locked_formula_count: 8,
    locked_formula_ids: ["F1", "F4", "F7", "F11", "F12", "F18", "F19", "F22"],
  },
});
assert.equal(observed.state, "observed");
assert.equal(observed.label, "8");
assert.equal(observed.count, 8);
assert.deepEqual(observed.ids, [
  "F1",
  "F4",
  "F7",
  "F11",
  "F12",
  "F18",
  "F19",
  "F22",
]);

const genome = policy.classifyHonestPayload({ locked_formula_count: 25 });
assert.equal(genome.state, "unavailable");
assert.equal(genome.label, "N/A");
assert.equal(genome.count, null);

const missing = policy.classifyHonestPayload({ footer: "Doctrine v11" });
assert.equal(missing.label, "N/A");

const failed = policy.classifyFetchFailure("HONEST_REQUEST_FAILED");
assert.equal(failed.label, "UNAVAILABLE");
assert.equal(failed.state, "unavailable");
assert.equal(failed.count, null);

function jsonResponse(payload, url) {
  return {
    ok: true,
    redirected: false,
    url: url || policy.HONEST_URL,
    json: async () => payload,
  };
}

const liveEight = await policy.fetchHonestManifest({
  url: policy.HONEST_URL,
  fetch: async (url) => {
    assert.equal(url, policy.HONEST_URL);
    return jsonResponse({ locked_formula_count: 8 });
  },
});
assert.equal(liveEight.label, "8");
assert.equal(liveEight.count, 8);

const liveWrong = await policy.fetchHonestManifest({
  fetch: async () => jsonResponse({ locked_formula_count: 25 }),
});
assert.equal(liveWrong.label, "N/A");

const liveFail = await policy.fetchHonestManifest({
  fetch: async () => {
    throw new Error("network");
  },
});
assert.equal(liveFail.label, "UNAVAILABLE");

const httpFail = await policy.fetchHonestManifest({
  fetch: async () => ({
    ok: false,
    redirected: false,
    url: policy.HONEST_URL,
    json: async () => ({ locked_formula_count: 8 }),
  }),
});
assert.equal(httpFail.label, "UNAVAILABLE");

const redirectFail = await policy.fetchHonestManifest({
  fetch: async () => ({
    ok: true,
    redirected: true,
    url: policy.HONEST_URL,
    json: async () => ({ locked_formula_count: 8 }),
  }),
});
assert.equal(redirectFail.label, "UNAVAILABLE");

const noFetch = await policy.fetchHonestManifest({});
assert.equal(noFetch.label, "UNAVAILABLE");

function chip(id) {
  const count = { textContent: "N/A", nodeType: 1 };
  const ids = { textContent: "", nodeType: 1 };
  const el = {
    id,
    dataset: { kernelChip: "locked-proven", state: "unavailable" },
    querySelector(selector) {
      if (selector === "[data-kernel-count]" || selector === "b") return count;
      if (selector === "[data-kernel-ids]") return ids;
      return null;
    },
    closest() {
      return el;
    },
    _count: count,
    _ids: ids,
  };
  return el;
}

const hero = chip("cnt-locked");
policy.paintChip(hero, observed);
assert.equal(hero._count.textContent, "8");
assert.equal(hero.dataset.state, "observed");
assert.equal(hero._ids.textContent, " {F1, F4, F7, F11, F12, F18, F19, F22}");

policy.paintChip(hero, genome);
assert.equal(hero._count.textContent, "N/A");
assert.equal(hero.dataset.state, "unavailable");
assert.equal(hero._ids.textContent, "");

policy.paintChip(hero, failed);
assert.equal(hero._count.textContent, "UNAVAILABLE");

const source = require("node:fs").readFileSync(
  new URL("./honest_kernel_bind.js", import.meta.url),
  "utf8",
);
assert.doesNotMatch(source, /\?\?\s*8\b/);
assert.doesNotMatch(source, /\|\|\s*8\b/);
assert.doesNotMatch(source, /exactly\s+8/i);
assert.match(source, /\/api\/a11oy\/v1\/honest/);
assert.match(source, /locked_formula_count/);
assert.match(source, /value === 8/);
assert.match(source, /N\/A/);
assert.match(source, /factory\(root\)/);
assert.match(source, /function \(root\)/);
assert.match(source, /mode:\s*["']cors["']/);
assert.doesNotMatch(source, /var policy = factory\(\)/);

import vm from "node:vm";
const autoCount = { textContent: "N/A" };
const autoChip = {
  dataset: { state: "unavailable" },
  querySelector(selector) {
    if (selector === "[data-kernel-count]" || selector === "b") return autoCount;
    if (selector === "[data-kernel-ids]") return { textContent: "" };
    return null;
  },
};
const sandbox = {
  document: {
    readyState: "complete",
    addEventListener() {},
    querySelectorAll(selector) {
      return selector === "[data-kernel-chip]" ? [autoChip] : [];
    },
    getElementById() {
      return autoChip;
    },
  },
  fetch: async (url) => {
    assert.equal(url, policy.HONEST_URL);
    return {
      ok: true,
      redirected: false,
      url,
      json: async () => ({ doctrine_lock: { locked_formula_count: 8 } }),
    };
  },
  setTimeout,
  clearTimeout,
  AbortController,
  URL,
  Promise,
  Object,
  Array,
  TypeError,
  Error,
  Math,
  isFinite,
  console,
};
sandbox.globalThis = sandbox;
vm.runInNewContext(source, sandbox);
await new Promise((resolve) => setTimeout(resolve, 50));
assert.equal(sandbox.A11oyHonestKernelBind.HONEST_FIELD, "locked_formula_count");
assert.equal(autoCount.textContent, "8");
assert.equal(autoChip.dataset.state, "observed");

console.log(
  "OK: kernel chips bind /honest locked_formula_count === 8 or N/A; fetch failure is UNAVAILABLE.",
);
